features/48-user-profile-discovery/user-profile.md

```diff
@@ -0,0 +1,258 @@
+
+- `container.discoverUserByUsername(username)`
+- `container.discoverUserByUsernames(usernames)`
+- `container.discoverUserByEmail(email)`
+- `container.discoverUserByEmails(emails)`
+
+## New Sign Up and Log In API
+
+### Sign Up API
+
+The new sign up API is proposed to enable: 1) sign up with custom data;
+2) update user profile during sign up. The new sign up API will perform:
+
+1. Discover user with the authentication data
+1. Sign up with the authentication data and password, if no user is found
+1. Update user record as profile
+
+The function signature will be `container.signup(authData, password, profile)`
+while `authData` is used to discover user. The following sign up API will be
+kept as helper methods:
+
+- `container.signupWithUsername(username, password)`
+- `container.signupWithEmail(email, password)`
+- `container.signupAnonymously()`
+
+### Log In API
+
+The similar signature is also proposed to log in API to enable log in with
+custom data: `container.login(authData, password)`.  The new log in API will
+perform:
+
+1. Discover user with the authentication data
+1. Log in with the authentication data and password, if only one user is found
+
+The current log in API will be kept as helper methods:
+
+- `container.loginWithUsername(username, password)`
+- `container.loginWithEmail(email, password)`
+
+### Authentication Record Configurations
+
+Since Skygear Server does not have a way for developers to set the uniqueness
+constraint of record fields (except writing the `before_save` hook of the
+record), configurations are proposed for authentication record.
+
+Authentication Record Configurations should contains: 1) Authentication Record
+Type; 2) Authentication Record Fields. They are suggested to be provided by
+environment variables during implementation:
+
+- `AUTH_RECORD_TYPE`: the record type for authentication, default: `user`
+- `AUTH_RECORD_FIELDS`: the record field names of the record for
+  authentication. The format will be a comma separated list while each list
+  item is either a field name or another comma separated list wrapped with a
+  pair of brackets. (e.g. `slug,(username,role),(email,role)`). The default
+  value would be `username,email`.
+
+The authentication record field configuration defines the following behavior:
+
+- the list of unique constraints on the corresponding table of the
+  authentication record.
+- the list of available authentication data for sign up / log in on non
+  development mode. For example, if `AUTH_RECORD_FIELDS` is set to
+  `slug,(username,role)`, only the following formats of authentication data
+  are accepted on non development mode: 1) `slug` only; 2) `username` together
+  with `role`.
+
+# Sample Codes for Use Cases
+
+## User discovery API
+
+The following sample code shows how to use user discovery API for different
+purposes:
+
+```js
+const skygear = require('skygear');
+
+// Use Case 1: discover user by username
+skygear.discoverUserByUsername('cheungpat')
```

Comment:
On the Design Principle, should be if find one users / find a list of users is "one thing" or "different things"?  I think it might justify for some use cases which need a list of users quickly.

However I'm not sure if the word "discover" is good as mentioned above.